iroh/src/magicsock/relay_actor.rs

```diff
@@ -788,6 +789,105 @@ impl ActiveRelayActor {
     }
 }
 
+/// Forward pending received packets to their queues.
+///
+/// If `maybe_pending` is not empty, this will wait for the path the last received item
+/// is blocked on (via [`PendingRecv::blocked_on`]) to become unblocked. It will then forward
+/// the pending items, until a queue is blocked again. In that case, the remaining items will
+/// be put back into `maybe_pending`. If all items could be sent, `maybe_pending` will be set
+/// to `None`.
+///
+/// This function is cancellation-safe: If the future is dropped at any point, all items are guaranteed
+/// to either be sent into their respective queues, or are still in `maybe_pending`.
+async fn forward_pending(
+    maybe_pending: &mut Option<PendingRecv>,
+    relay_datagrams_recv: &RelayDatagramRecvQueue,
+    relay_disco_recv: &mut mpsc::Sender<RelayDiscoMessage>,
+) {
+    // We take a mutable reference onto the inner value.
+    // we're not `take`ing it here, because this would make the function not cancellation safe.
+    let Some(ref mut pending) = maybe_pending else {
+        return;
+    };
+    let disco_permit = match pending.blocked_on {
+        RecvPath::Data => {
+            std::future::poll_fn(|cx| relay_datagrams_recv.poll_send_ready(cx))
+                .await
+                .ok();
+            None
+        }
+        RecvPath::Disco => {
+            let Ok(permit) = relay_disco_recv.clone().reserve_owned().await else {
+                return;
+            };
+            Some(permit)
+        }
+    };
+    // We now take the inner value by value. it is cancellation safe here because
+    // no further `await`s occur after here.
+    // The unwrap is guaranteed to be safe because we checked above that it is not none.
+    #[allow(clippy::unwrap_used, reason = "checked above")]
+    let pending = maybe_pending.take().unwrap();
```

Comment:
We use `.expect()` for cases like this, I thought `.unwrap()` was denied by some linter even?